### ISSUE TITLE
Bad Performance of text-generation with sampling algo

### ISSUE THREAD

**Issue by ZhaiFeiyue** [huggingface/optimum-habana#252]
### System Info

```shell
Ubuntu 20.04
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

with `do_sample`, the throughput is 450 tps
```bash
python run_generation.py --model_name_or_path gpt2  --batch_size 1 --max_new_tokens 100 --use_hpu_graphs --prompt "I am a student from" --do_sample
```

without `do_sample`, the throughput is 900tps

```bash
python run_generation.py --model_name_or_path gpt2  --batch_size 1 --max_new_tokens 100 --use_hpu_graphs --prompt "I am a student from"
```

### Expected behavior

decent performance with `do_sample` algo

**Comment by ZhaiFeiyue** [comment]
@regisss  you have asked me before about the bad performance when you runing text-generation with sampling algo, after profiling, I think the issue is bad pipelining of host and device, see the `to_cpu` part with red rectangle.
![image](https://github.com/huggingface/optimum-habana/assets/80079571/42dbed11-3b77-4111-b8d0-885299241d68)

it block host until device finished. so what I think is try to remove it by introducing `check early stop every N steps`.

**Comment by regisss** [comment]
Interesting! Not sure what you mean by "check early stop every N steps" though, could you explain it a bit more?

**Comment by ZhaiFeiyue** [comment]
from the profile we could see there is a `to_cpu` and `to_hpu` and after investigation, I found it is because tensor.prod is a fallback op in HPU side refer code
![image](https://github.com/huggingface/optimum-habana/assets/80079571/db697e68-8c20-4525-9e35-30d35ae63112)

when there is a CPU fallback op, which means we must stop here and wait the tensor from device ready.
the above part is used for early stop checking, so a straight  forward solution is `checking early stop every 5 or 10 step`, not calling above part code every step.

But I have not seen this issue before, and after check the code I found that we are using a old transformers, with which the implementation is
`unfinished_sequences = unfinished_sequences.mul((sum(next_tokens != i for i in eos_token_id)).long())`

**Comment by ZhaiFeiyue** [comment]
@regisss  seems much better now
before
![image](https://github.com/huggingface/optimum-habana/assets/80079571/11400b2d-77fb-4163-a8b5-58672eb60f51)

now is 
![image](https://github.com/huggingface/optimum-habana/assets/80079571/ba5e5ce9-b7cc-4e83-8b90-80cef49a4e51)

cost of one step on device is reduced from 2.2ms to 1.9ms
but there is still a waiting
![image](https://github.com/huggingface/optimum-habana/assets/80079571/f3f17b12-dbf9-4b85-b4f9-cbdd16227ff9)

I think it is comes from here
![image](https://github.com/huggingface/optimum-habana/assets/80079571/d3b9d19f-989d-4621-8a02-6f440f26ea6f)

how about we adding a feature that `check early stop every N step`?

**Comment by regisss** [comment]
I took a closer look and I have a few questions/remarks:
- We should see the same behavior with `greedy_search` no? Since early stopping is managed the same way: https://github.com/huggingface/optimum-habana/blob/2cb7646bdaa41e88e05ef2bea18fce43f2faf0a8/optimum/habana/transformers/generation/utils.py#L1173
- You mentioned that this was not happening with the previous implementation. Maybe we can just revert back to
  ```py
  unfinished_sequences = unfinished_sequences.mul((sum(next_tokens != i for i in eos_token_id)).long())
  ```
  what do you think?
- We use early stopping in eager mode only, so a quick and simple fix would be to change
  ```py
  if eos_token_id_tensor is not None:
      unfinished_sequences = unfinished_sequences.mul(
          next_tokens.tile(eos_token_id_tensor.shape[0], 1).ne(eos_token_id_tensor.unsqueeze(1)).prod(dim=0)
      )
  ```
  to
  ```py
  if not lazy_mode and eos_token_id_tensor is not None:
      unfinished_sequences = unfinished_sequences.mul(
          next_tokens.tile(eos_token_id_tensor.shape[0], 1).ne(eos_token_id_tensor.unsqueeze(1)).prod(dim=0)
      )
  ```
  if I understood correctly.

**Comment by ZhaiFeiyue** [comment]
yes @regisss , you are right, but not enough
the following code will still block and wait
![image](https://github.com/huggingface/optimum-habana/assets/80079571/7cd37166-4b90-4051-9d19-7e6ff72cf978)

what is the `ignore_eos`, after debugging I found that in `greedy_search`, the `ignore_eos` is True, so we will skip the checking

**Comment by regisss** [comment]
Good catch, we don't use `ignore_eos` when doing sampling. So I guess we can just use it the same way as in `greedy_search`, and that should solve this issue?

**Comment by ZhaiFeiyue** [comment]
how about the following changes
![image](https://github.com/huggingface/optimum-habana/assets/80079571/beef65b9-37e5-474d-a72f-5c7692d4ed99)

checking `azy_mode` first and if True, python won't check `unfinished_sequences.max() == 0` and there is no wait
and perf is
![image](https://github.com/huggingface/optimum-habana/assets/80079571/ebc2baf7-2b69-4a08-bee4-778d3668aef2)

**Comment by ZhaiFeiyue** [comment]
closed since PR #254 fixed
